working/0723-static-extensions/feature-specification-variant1.md

```diff
@@ -0,0 +1,467 @@
+declaration named `m` were found. *They would necessarily be declared in
+static extensions.*
+
+Otherwise, the invocation is resolved to the given static member
+declaration in a static extension named `Ej`, and the invocation is treated
+as `Ej.m()` *(this is an explicitly resolved invocation, which is specified
+above)*.
+
+#### The instantiated constructor return type of a static extension
+
+We associate a static extension declaration _D_ named `E` with formal type
+parameters `X1 extends B1 .. Xs extends Bs` and an actual type argument
+list `T1 .. Ts` with a type known as the _instantiated constructor return
+type of_ _D_ _with type arguments_ `T1 .. Ts`.
+
+When a static extension declaration _D_ named `E` has an on-clause which
+denotes a non-generic class `C`, the instantiated constructor return type
+is `C`, for any list of actual type arguments.
+
+*It is not very useful to declare a type parameter of a static extension
+which isn't used in the constructor return type, because it can only be
+passed in an explicitly resolved constructor invocation, e.g.,
+`E<int>.C(42)`. In all other invocations, the value of such type variables
+is determined by instantiation to bound. In any case, the type parameters
+are always ignored by static member declarations, they are only relevant to
+constructors.*
+
+When a static extension declaration _D_ has no formal type parameters, and
+it has an on-type `C<S1 .. Sk>`, the instantiated constructor return type
+of _D_ is `C<S1 .. Sk>`. *In this case the on-type is a fixed type (also
+known as a ground type), e.g., `List<int>`. This implies that the
+constructor return type of D is the same for every call site.*
+
+Finally we have the general case: Consider a static extension declaration
+_D_ named `E` with formal type parameters `X1 extends B1 .. Xs extends Bs`
+and a constructor return type `C<S1 .. Sk>`. With actual type arguments
+`T1 .. Ts`, the instantiated constructor return type of _D_ with type
+arguments `T1 .. Ts` is `[T1/X1 .. Ts/Xs]C<S1 .. Sk>`.
+
+#### Explicit invocation of a constructor in a static extension
+
+Explicit constructor invocations are similar to static member invocations,
+but they need more detailed rules because they can use the formal type
+parameters declared by a static extension.
+
+An _explicitly resolved invocation_ of a constructor named `C.name` in a
+static extension declaration _D_ named `E` with `s` type parameters and
+on-class `C` can be expressed as `E<S1 .. Ss>.C.name(args)`,
+`E.C<U1 .. Uk>.name(args)`, or `E<S1 .. Ss>.C<U1 .. Uk>.name(args)`
+(and similarly for a constructor named `C` using `E<S1 .. Ss>.C(args)`,
+etc).
+
+A compile-time error occurs if the type arguments passed to `E` violate the
+declared bounds. A compile-time error occurs if no type arguments are
+passed to `E`, and type arguments `U1 .. Uk` are passed to `C`, but no list
+of actual type arguments for the type variables of `E` can be found such
```

Comment:
"can be found" how?  That is, we must specify what mechanism is used to find it (since otherwise "I can't find it, sorry" is always a valid answer).

Reply:
Right, but that's true for every task which is loosely described as type inference. We need an algorithm (which is known to be terminating, etc.) in order to get an implementation, but at this point I wanted to specify the constraints that such an algorithm must satisfy.

I believe we should use an algorithm which is different from the one which is otherwise used in order to infer types in dart. That is, I don't think we can directly use something like `E_Map_bar` in https://github.com/dart-lang/language/issues/4053. I think we need to use equality constraints rather than subtype constraints, because we can't have something like `new List<num>.qux()` that returns a `List<int>` or a `List<Object>` (we _could_ specify that the static type of `new List<num>.qux()` is `List<Object>` and it could yield a `List<Object>`, but I think we shouldn't).

In other words, the type arguments for the extension will fully determine the type arguments for the class. If we already have the type arguments for the class and not for the extension then I believe we must find type arguments for the extension such that substitution into the on-type yields _exactly_ the type arguments for the class that we already have (say, because a developer wrote them explicitly).

Equality constraints would basically just have the structural steps, so it should be _a lot_ simpler than general type inference.

```dart
static extension E<X, Y> on Map<String, X> {
  factory Map.name(Y y, X Function(Y) fun) => {y.toString(): fun(y)};
}

void main() {
  var map = Map<String, double>.name(14, (i) => i + 0.5);
}
```

In this case we resolve the invocation of `Map.name` as an invocation of `E.Map.name`, say, because no other static extension declares a static member named `name` or a constructor named `Map.name`.

Then we generate constraints: `X == double`. Next, we infer `E<double, _>.C<String, double>.name` like

```dart
Map<String, double> E_Map_name<Y>(Y y, double Function(Y) fun) =>
    throw "Just for inference";

void main() {
  var map = E_Map_name(14, (i) => i + 0.5); // Infers `E_Map_name<int>(...)`.
}
```

This is the case where the actual type arguments to the class have been specified explicitly. A similar case arises when there is a context type that determines the value of type variables of `E`:

```dart
void main() {
  Map<Object, double> map = Map.name(14, (i) => i + 0.5);
}
```

In this case I'm proposing that we should use the context type. This binds `X` to `double`, which means that we have an instantiated constructor return type of `Map<String, double>`, and we get the value `int` for `Y` again.

Reply:
This discussion is now handled via https://github.com/dart-lang/language/issues/4053.